### PR TITLE
Metrics as autosummary tables

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Reformat metric docs to use autosummary.
Requested by @tchaton on slack :]

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+the functional metrics are the way to go. However, if you are looking for the
+best integration and user experience, please consider also to use the class interface.
+
+Classification [func]
+---------------------
+
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+    :template: classtemplate.rst
```
Comment (ydcjeff) [inline docs/source/metrics.rst:263]:
```suggestion
    :template: functiontemplate.rst
```

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+    to_categorical
+    to_onehot
+
+Regression [func]
+-----------------
+
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+    :template: classtemplate.rst
```
Comment (ydcjeff) [inline docs/source/metrics.rst:292]:
```suggestion
    :template: functiontemplate.rst
```

File: docs/source/metrics.rst [hunk 3]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+
+NLP [func]
+----------
 
-bleu_score [func]
-~~~~~~~~~~~~~~~~~
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+    :template: classtemplate.rst
```
Comment (ydcjeff) [inline docs/source/metrics.rst:307]:
```suggestion
    :template: functiontemplate.rst
```

**Review by Borda — COMMENTED** [review]
can we add screenshot of the new one?

**Comment by SkafteNicki** [comment]
Untitled

**Review by ananyahjha93 — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Hey @SkafteNicki, looks great ! Thanks !

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -150,8 +150,13 @@ To implement your custom metric, subclass the base ``Metric`` class and implemen
 All you need to do is call ``add_state`` correctly to implement a custom metric with DDP.
 ``reset()`` is called on metric state variables added using ``add_state()``.
 
-To see how metric states are synchronized across distributed processes, refer to ``add_state()`` docs
-from the base ``Metric`` class.
+.. note::
+
+    ``reset()`` is automatically called whenever ``compute()`` is called. Therefore
+    you should very rearly have to call ``reset()`` yourself.
```
Comment (rohitgr7) [inline docs/source/metrics.rst:156]:
```suggestion
    you should very rarely have to call ``reset()`` yourself.
```

File: docs/source/metrics.rst [hunk 5]
```diff
@@ -196,299 +201,120 @@ Metric API
...
-
-.. autoclass:: pytorch_lightning.metrics.regression.SSIM
-    :noindex:
 
-******************
-Functional Metrics
-******************
+****************
+Class metrics
+****************
```
Comment (rohitgr7) [inline docs/source/metrics.rst:207]:
```suggestion
*************
Class metrics
*************
```

File: docs/source/metrics.rst [hunk 6]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+    :template: classtemplate.rst
 
+    Accuracy
+    Precision
+    Recall
+    Fbeta
+    ConfusionMatrix
```
Comment (rohitgr7) [inline docs/source/metrics.rst:224]:
```suggestion
    Accuracy
    ConfusionMatrix
    Fbeta
    Precision
    Recall
```
alphabetical order maybe?

File: docs/source/metrics.rst [hunk 7]
```diff
@@ -196,299 +201,120 @@ Metric API
...
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-.. autofunction:: pytorch_lightning.metrics.functional.mean_squared_log_error
-    :noindex:
+    MeanSquaredError
+    MeanAbsoluteError
+    MeanSquaredLogError
+    ExplainedVariance
+    PSNR
+    SSIM
```
Comment (rohitgr7) [inline docs/source/metrics.rst:239]:
```suggestion
    ExplainedVariance
    MeanAbsoluteError
    MeanSquaredLogError
    MeanSquaredError
    PSNR
    SSIM
```
alphabetical order maybe?

File: docs/source/metrics.rst [hunk 8]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+.. currentmodule:: pytorch_lightning.metrics.functional
+
+The functional metrics follow the simple paradigm: input in, output out. This means,
+they don't provide any advanced mechanisms for syncing across DDP nodes or aggregation
+over batches. They simply compute the metric value based on the given inputs.
+
+Also the integration within other parts of PyTorch Lightning will never be as tight
+as with the class-based interface. If you look for just computing the values,
+the functional metrics are the way to go. However, if you are looking for the
+best integration and user experience, please consider also to use the class interface.
```
Comment (rohitgr7) [inline docs/source/metrics.rst:255]:
```suggestion
best integration and user experience, please consider to use the class interface.
```

File: docs/source/metrics.rst [hunk 9]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+    :toctree: generated
+    :nosignatures:
+    :template: functiontemplate.rst
+
+    explained_variance
+    mean_absolute_error
+    mean_squared_error
+    psnr
+    mean_squared_log_error
+    ssim
```
Comment (rohitgr7) [inline docs/source/metrics.rst:299]:
```suggestion
    explained_variance
    mean_absolute_error
    mean_squared_log_error
    mean_squared_error
    psnr
    ssim
```

File: docs/source/metrics.rst [hunk 10]
```diff
@@ -196,299 +201,120 @@ Metric API
...
+    multiclass_roc
+    precision
+    precision_recall
+    precision_recall_curve
+    recall
+    roc
+    stat_scores
+    stat_scores_multiple_classes
+    to_categorical
+    to_onehot
```
Comment (rohitgr7) [inline docs/source/metrics.rst:284]:
```suggestion
    accuracy
    auc
    auroc
    average_precision
    confusion_matrix
    dice_score
    f1_score
    fbeta_score
    iou
    multiclass_auroc
    multiclass_roc
    precision
    precision_recall
    precision_recall_curve
    recall
    roc
    stat_scores
    stat_scores_multiple_classes
    to_categorical
    to_onehot
```

**Review by rohitgr7 — APPROVED** [review]
LGTM. Just some comments.

**Review by williamFalcon — CHANGES_REQUESTED** [review]
before merging can i clarify something?

the metric summary tables remove the menus on the right that make it super easy to search for stuff without having to scroll endlessly through the page. 

is that menu gone now?

if so, can we undo all these summary tables? they make quickly scanning to see what’s available impossible. 

to see how useful this menu is, try finding any trainer flag. now imagine if the menu ok the right didn’t exist and you instead had to scroll the full page just to see what flags are available. 

remember that our docs are as much about discovery as they are about referencing

**Comment by SkafteNicki** [comment]
The menu is not all gone, it still expands to the individual types of metrics:
Untitled2
If the user is aware what type of metric they are looking for (classification, regression, ect) I would argue that it requires much less scrolling now.

**Comment by williamFalcon** [comment]
right. this is exactly what i think is not a great user experience. 

we now remove the ability for the user to discover what is available by just glancing at the right since all the metrics are gone from the menu. 

i also want to remove these tables from the rest of the docs. 

this approach assumes people know what they are looking for but makes it impossible for people who are just browsing to see what’s available... and actually the menu on the right showing the list of metrics is much much faster to find the metric you are looking for. 

@tchaton, @edenafek can we remove these tables from the docs and do it exactly like we do for trainer flags?

**Comment by tchaton** [comment]
Hey @SkafteNicki, 

After chatting with Will, 
we thought it is better to keep the Metric as they are, but make the right slide panel bigger instead and take inspiration from Trainer.

I deeply apologise as you invested time working on this and it won't get integrated.
I will be very careful when suggesting this kind of changes in the future.

Best regards,
Thomas Chaton.

**Comment by SkafteNicki** [comment]
@tchaton it is completely fine, I did not spend that much time on this PR :]
Closing this pr.

**Pull request closed without merging**